rusoto/credential/src/environment.rs

```diff
@@ -52,24 +43,98 @@ const E_INVALID_EXPIRATION: &str = "Invalid AWS_CREDENTIAL_EXPIRATION in environ
 /// env::set_var("AWS_CREDENTIAL_EXPIRATION", "2018-04-21T01:13:02Z");
-/// let creds = EnvironmentProvider.credentials().wait().unwrap();
+/// let creds = EnvironmentProvider::default().credentials().wait().unwrap();
 /// assert_eq!(creds.expires_at().unwrap().to_rfc3339(), "2018-04-21T01:13:02+00:00");
 /// # }
 /// ```
-#[derive(Debug)]
-pub struct EnvironmentProvider;
+#[derive(Debug, Clone)]
+pub struct EnvironmentProvider {
+    prefix: String,
+}
+
+impl Default for EnvironmentProvider {
+    fn default() -> Self {
+        EnvironmentProvider {
+            prefix: "AWS".to_owned(),
+        }
+    }
+}
+
+impl EnvironmentProvider {
+    /// Create an EnvironmentProvider with a non-standard variable prefix.
+    ///
+    /// ```rust
+    /// # extern crate futures;
+    /// # extern crate rusoto_credential;
+    /// #
+    /// # fn main() {
+    /// use futures::future::Future;
+    /// use rusoto_credential::{EnvironmentProvider, ProvideAwsCredentials};
+    /// use std::env;
+    ///
+    /// env::set_var("MYAPP_ACCESS_KEY_ID", "ANTN35UAENTS5UIAEATD");
+    /// env::set_var("MYAPP_SECRET_ACCESS_KEY", "TtnuieannGt2rGuie2t8Tt7urarg5nauedRndrur");
+    /// env::set_var("MYAPP_SESSION_TOKEN", "DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir");
+    ///
+    /// let creds = EnvironmentProvider::with_prefix("MYAPP").credentials().wait().unwrap();
+    ///
+    /// assert_eq!(creds.aws_access_key_id(), "ANTN35UAENTS5UIAEATD");
+    /// assert_eq!(creds.aws_secret_access_key(), "TtnuieannGt2rGuie2t8Tt7urarg5nauedRndrur");
+    /// assert_eq!(creds.token(), &Some("DfnGs8Td4rT8r4srxAg6Td4rT8r4srxAg6GtkTir".to_string()));
+    /// assert!(creds.expires_at().is_none()); // doesn't expire
+    ///
+    /// env::set_var("MYAPP_CREDENTIAL_EXPIRATION", "2018-04-21T01:13:02Z");
+    /// let creds = EnvironmentProvider::with_prefix("MYAPP").credentials().wait().unwrap();
+    /// assert_eq!(creds.expires_at().unwrap().to_rfc3339(), "2018-04-21T01:13:02+00:00");
+    /// # }
+    /// ```
+    pub fn with_prefix(prefix: &str) -> Self {
+        EnvironmentProvider {
+            prefix: prefix.to_owned(),
+        }
+    }
+}
+
+/// A private trait for building the environment variable names based
+/// on a provided prefix. Pulled out of the base `EnvironmentProvider`
+/// to keep separate from the public functionality.
+trait EnvironmentVariableProvider {
+    fn prefix(self: &Self) -> String;
+
+    fn access_key_id_var(self: &Self) -> String {
+        format!("{}_ACCESS_KEY_ID", self.prefix())
+    }
+
+    fn secret_access_key_var(self: &Self) -> String {
+        format!("{}_SECRET_ACCESS_KEY", self.prefix())
+    }
+
+    fn session_token_var(self: &Self) -> String {
+        format!("{}_SESSION_TOKEN", self.prefix())
+    }
+
+    fn credential_expiration_var(self: &Self) -> String {
+        format!("{}_CREDENTIAL_EXPIRATION", self.prefix())
+    }
+}
+
+impl EnvironmentVariableProvider for EnvironmentProvider {
+    fn prefix(self: &Self) -> String {
```

Comment:
Given that the prefix is used as part of a format string, there doesn't seem to be a lot of value in returning it in an owned form. What do you think about `prefix` returning `&str` instead? That would also prevent the extra clone that you need to do here.

Reply:
Good call, I'll make the change.